### PR TITLE
Reworking client to be better future proof.

### PR DESCRIPTION
**Pull request by n3wscott**
Major Changes:

[x] Removed `client/[http, nats]` helper classes. It forced the loss of type safety for options and gives no hints on what the new method accepts. Better to be explicit. 
[x] Adds `client.NewDefault()`, which takes no options. This can be used for the standard cases of sending and receiving for normal HTTP situations.
[x] `client.StartReceiver` is now a blocking call.
[x] Add mutex around StartReceiver.

Signed-off-by: Scott Nichols <nicholss@google.com>

### CODE REVIEW

**Comment by n3wscott** [comment]
https://github.com/cloudevents/sdk-go/pull/68 should go in first, there will be come conflicts.

File: pkg/cloudevents/transport/http/transport.go [hunk 1]
```diff
@@ -171,26 +176,28 @@ func (t *Transport) StartReceiver(ctx context.Context) error {
 		}
 		t.realPort = listener.Addr().(*net.TCPAddr).Port
 
-		go func() {
-			log.Print(t.server.Serve(listener))
-			t.server = nil
+		// Shutdown
+		defer func() {
 			t.realPort = 0
+			t.server.Close()
```
Comment (grantr) [inline pkg/cloudevents/transport/http/transport.go]: Might be better to use graceful `Shutdown(ctx)` here, but I'm not sure how to choose a shutdown timeout.

File: pkg/cloudevents/transport/http/transport.go [hunk 2]
```diff
@@ -171,26 +176,28 @@ func (t *Transport) StartReceiver(ctx context.Context) error {
...
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errChan:
+			return err
 		}
+
 	}
-	return nil
+	return fmt.Errorf("http server already started")
```
Comment (grantr) [inline pkg/cloudevents/transport/http/transport.go]: Now that there's a mutex around the entire (blocking) method, can this ever be reached?

Comment (n3wscott) [reply pkg/cloudevents/transport/http/transport.go]: nope!

File: cmd/samples/http/receiver/main.go [hunk 3]
```diff
@@ -41,7 +40,7 @@ func gotEvent(ctx context.Context, event cloudevents.Event) error {
 	}
 	fmt.Printf("Got Data: %+v\n", data)
 
-	fmt.Printf("Got Transport Context: %+v\n", http.TransportContextFrom(ctx))
+	fmt.Printf("Got Transport Context: %+v\n", cloudeventshttp.TransportContextFrom(ctx))
```
Comment (markpeek) [inline cmd/samples/http/receiver/main.go]: other samples are using cehttp...do we want consistency in this naming?

Comment (n3wscott) [reply cmd/samples/http/receiver/main.go]: I have not made a choice on which is best.

File: pkg/cloudevents/transport/http/transport.go [hunk 4]
```diff
@@ -154,50 +154,47 @@ func (t *Transport) StartReceiver(ctx context.Context) error {
...
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- t.server.Serve(listener)
+	}()
+
+	// wait for the server to return or ctx.Done().
+	select {
+	case <-ctx.Done():
+		// Try a gracefully shutdown.
+		return t.server.Shutdown(context.Background())
```
Comment (grantr) [inline pkg/cloudevents/transport/http/transport.go]: When called with the background context, `Shutdown()` can block indefinitely. Is that what you want here?

**Pull request merged**